src/interface/variable_pack.hpp

```diff
@@ -0,0 +1,234 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//=======================================================================================
+#ifndef INTERFACE_VARIABLE_PACK_HPP_
+#define INTERFACE_VARIABLE_PACK_HPP_
+
+#include <array>
+#include <forward_list>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "interface/metadata.hpp"
+#include "interface/variable.hpp"
+
+namespace parthenon {
+
+// some convenience aliases
+namespace vpack_types {
+template <typename T>
+using VarList = std::forward_list<std::shared_ptr<CellVariable<T>>>;
+using IndexPair = std::pair<int, int>;
+using StringPair = std::pair<std::vector<std::string>, std::vector<std::string>>;
+} // namespace vpack_types
+
+using PackIndexMap = std::map<std::string, vpack_types::IndexPair>;
+template <typename T>
+using ViewOfParArrays = ParArrayND<ParArray3D<T>>;
+
+// Try to keep these Variable*Pack classes as lightweight as possible.
+// They go to the device.
+template <typename T>
+class VariablePack {
+ public:
+  VariablePack() = default;
+  VariablePack(const ViewOfParArrays<T> view, const std::array<int, 4> dims)
+      : v_(view), dims_(dims) {}
+  KOKKOS_FORCEINLINE_FUNCTION
+  ParArray3D<T> &operator()(const int n) const { return v_(n); }
+  KOKKOS_FORCEINLINE_FUNCTION
+  T &operator()(const int n, const int k, const int j, const int i) const {
+    return v_(n)(k, j, i);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetDim(const int i) {
+    assert(i > 0 && i < 5);
+    return dims_[i - 1];
+  }
+
+ protected:
+  ViewOfParArrays<T> v_;
+  std::array<int, 4> dims_;
```

Comment:
If this goes to the device, do we need to use a `Kokkos::array` (or whatever it is)?

Reply:
`std::array` appears to work in cuda lambda captures. A C-style array does not, hence the `stdlib` use.